### PR TITLE
feat(azure): add support for managed identities

### PR DESCRIPTION
**Pull request by wallyworld**
Azure controllers can now be configured to use user managed identities via the instance-role constraint.

Some small refactoring of the resource group set up was needed since the group now needs to exist for the managed identity to be created.

The feature works the similarly to what was done for AWS - we have an "instance-role" constraint attribute which can be used to specify the user managed identity to use. The constraint naming is a little unfortunate - it's somewhat AWS specific - but it's what we have and we don't want to introduce a new constraint attribute.

The user can create their own user managed identity or ask Juju to do it.

1. Juju creates managed identity
Here, the instance-role is set to "auto".
Juju will create a managed identity in the resource group used to host the controller model.
bootstrap:
`juju bootstrap azure --constraints="instance-role=auto" ...`

2. User supplied managed identity
This is more for advanced users who want to manually manage their Azure assets.
Here, the user needs to create their own managed identity with a role assignment conferring the following allowed actions:
```
"Microsoft.Compute/*",
"Microsoft.KeyVault/*",
"Microsoft.Network/*",
"Microsoft.Resources/*",
"Microsoft.Storage/*",
"Microsoft.ManagedIdentity/userAssignedIdentities/*"
```

The assignable scopes need to be compatible with the intended deployment, eg
```
"concat('/subscriptions/', subscription().subscriptionId)",
"resourceGroup().id"
```

The instance role can be located in a different group or even different subscription to that used to host the Juju controller.
The formats supported are:
`managedinstancename`
`groupname/managaedinstancename`
`subscription/groupname/managedinstancename`

bootstrap:
`juju bootstrap azure --constraints="instance-role=mymanagedinstance" ...`

## QA steps

Test various bootstrap scenarios. In each case, the controller credential can be inspected:
```
$ juju show-credential azure default
controller-credentials:
  azure:
    default:
      content:
        auth-type: instance-role
        validity-check: valid
        managed-identity: juju-controller-2e52ba20-068b-41f9-81d1-1ed9e84e52c6
        subscription-id: 2eebf55a-1e02-45d8-a299-02aed8aea00b
      models:
        controller: admin
```

Also deploy a charm with storage, eg postgresql.
Also run enable-ha for each and ensure all 3 controllers come up.

Scenario one - regression check
`juju bootstrap azure`
(here the controller credential will be a service principal)

Scenario two - juju managed resources
`juju bootstrap azure --constraints="instance-role=auto"`

Scenario three - custom resource group
First create the group, then
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=auto"`

Scenario four - manual managed identity
Create a managed identity in a resource group
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=mygroup/identityname"`

Scenario five - manual managed identity different group
Create a managed identity in a resource group
Create a different resource group
`juju bootstrap azure --config resource-group-name=mygroup --constraints="instance-role=othergroup/identityname"`

Bash CI test can be run - these cover the auto and custom managed identity case where the identity is created in some random group different to the group used for the controller.
You'll need to have the az client installed and be logged in to an Azure account.

`./main.sh -v -p azure azure_cloud`

## Documentation changes

@tmihoc the PR description covers the core concepts - we'll need to update the Azure provider doc along the same lines as would have been done for the AWS doc when this feature landed

## Links

**Jira card:** [JUJU-6206](https://warthogs.atlassian.net/browse/JUJU-6206)

[JUJU-6206]: https://warthogs.atlassian.net/browse/JUJU-6206?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hpidcock — COMMENTED** [review]
Could not validate `juju bootstrap azure --constraints="instance-role=auto"`, as discussed. I also think this needs an integration test.

**Comment by wallyworld** [comment]
> Could not validate `juju bootstrap azure --constraints="instance-role=auto"`, as discussed. I also think this needs an integration test.

We don't have integration tests for the AWS variant either. I was going to do a suite to cover both separately.

File: provider/azure/identity.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+// CreateAutoInstanceRole is responsible for setting up an instance role on
+// behalf of the user.
+// For Azure, this means creating a managed identity with the correct role definition
+// assigned to it.
+func (env *azureEnviron) CreateAutoInstanceRole(
+	ctx context.ProviderCallContext,
+	args environs.BootstrapParams,
+) (string, error) {
+	controllerUUID := args.ControllerConfig[controller.ControllerUUIDKey].(string)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply provider/azure/identity.go]: doh, yup

File: provider/azure/identity.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	controllerUUID string,
+) (*string, error) {
+	envTags := tags.ResourceTags(
+		names.ModelTag{},
+		names.NewControllerTag(controllerUUID),
+		env.config,
+	)
+
+	identityName := fmt.Sprintf("juju-controller-%s", controllerUUID)
+	roleName := fmt.Sprintf("juju-controllerrole-%s", controllerUUID)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply provider/azure/identity.go]: sure

**Comment by wallyworld** [comment]
> Could not validate `juju bootstrap azure --constraints="instance-role=auto"`, as discussed. I also think this needs an integration test.

Weird, this worked fine for me:
```
make go-install
juju bootstrap azure/westus --constraints="instance-role=auto" test --debug
```

**Comment by wallyworld** [comment]
> > Could not validate `juju bootstrap azure --constraints="instance-role=auto"`, as discussed. I also think this needs an integration test.
> 
> Weird, this worked fine for me:
> 
> ```
> make go-install
> juju bootstrap azure/westus --constraints="instance-role=auto" test --debug
> ```

Turns out I had a public-clouds.yaml with instance-role auth type in it. Removing that surfaced the issue which is now fixed.

File: tests/main.sh [hunk 3]
```diff
@@ -41,6 +41,7 @@ import_subdir_files includes
 # Please keep these in alphabetic order.
 TEST_NAMES="agents \
             appdata \
+            azure_cloud \
```
Comment (hpidcock) [inline tests/main.sh:44]: azure_cloud => cloud_azure

**Review by hpidcock — APPROVED** [review]
Thanks for the good shell tests. All the tests I did passed now.

File: tests/main.sh (hunk 3 above)
Comment (wallyworld) [reply tests/main.sh]: yup

**Comment by wallyworld** [comment]
Terraform test failure unrelated to this branch
```
 === RUN   TestAcc_ResourceApplication
    resource_application_test.go:30: Step 2/5 error: Error running apply: exit status 1
        
        Error: Client Error
        
          with juju_application.this,
          on terraform_plugin_test.tf line 16, in resource "juju_application" "this":
          16: resource "juju_application" "this" {
        
        Unable to create application, got error: cannot add application "test-app":
        charm "ch:amd64/jameinel-ubuntu-lite-10" not found
--- FAIL: TestAcc_ResourceApplication (10.64s)
```

**Comment by dparv** [comment]
Just coming to this one here, I tested and discussed with @wallyworld a bit, but this doesn't solve the problem that we are facing on Azure. The idea of using managed identity is to be able to bootstrap a controller without any user input - still using a service principal for this is a blocker, because deploying an application from the Azure marketplace with an ARM template doesn't allow the creation of an APP registration (only subscription owners can do that) and service principal to give access over a resource group for juju to manage. The managed resource group where the Canonical entity is granted full access is only created after the deployment starts, so this still blocks the automation of bootstrapping a controller in a 'no touch by the user' way.

**Comment by jnsgruk** [comment]
> Just coming to this one here, I tested and discussed with @wallyworld a bit, but this doesn't solve the problem that we are facing on Azure. The idea of using managed identity is to be able to bootstrap a controller without any user input - still using a service principal for this is a blocker, because deploying an application from the Azure marketplace with an ARM template doesn't allow the creation of an APP registration (only subscription owners can do that) and service principal to give access over a resource group for juju to manage. The managed resource group where the Canonical entity is granted full access is only created after the deployment starts, so this still blocks the automation of bootstrapping a controller in a 'no touch by the user' way.

So which managed identity were you expecting to use? Is there a managed identity attached to the machine where the juju client is running that could be used?

**Comment by dparv** [comment]
Yes, exactly, the managed identity attached to the VM. It has full access (Owner) over the managed resource group, created by the marketplace deployment, which allows the VM to build any resources in that  same resource group. There is a way to get oauth tokens https://learn.microsoft.com/en-us/azure/machine-learning/how-to-authenticate-online-endpoint that will get access over the resources they are allowed to (e.g. Owner will be able to do anything).

File: provider/azure/environ.go [hunk 4]
```diff
@@ -842,7 +873,30 @@ func (env *azureEnviron) createVirtualMachine(
...
 		DependsOn: vmDependsOn,
-	})
+	}
+	// For controllers, check to see if we need to assign a managed identity resource to the vm.
+	if instanceConfig.IsController() {
+		var managedIdentity string
+		if env.cloud.Credential.AuthType() == cloud.InstanceRoleAuthType {
+			// Add a new controller after bootstrap (enable-ha).
+			managedIdentity = env.cloud.Credential.Attributes()[credManagedIdentity]
+		} else if instanceConfig.Bootstrap.BootstrapMachineConstraints.HasInstanceRole() {
```
Comment (manadart) [inline provider/azure/environ.go:883]: This causes a panic when enabling HA. `IsController` is true, but we are not bootstrapping.

**Pull request merged**